finality-aleph/src/data_io.rs

```diff
@@ -1,22 +1,214 @@
+                    let keys : Vec<_> = self.dependent_messages.keys().cloned().collect();
+                    for block_hash in keys {
+                        if let Ok(Some(_)) = self.client.header(BlockId::Hash(block_hash)) {
+                            self.add_block(block_hash);
+                        }
+                    }
+                    maintenance_timeout = Delay::new(PERIODIC_MAINTENANCE_INTERVAL);
+                }
+                _ = &mut exit => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn forget_message(&mut self, message_id: MessageId) {
+        self.message_requirements.remove(&message_id);
+        if let Some(message) = self.pending_messages.remove(&message_id) {
+            for block_hash in message.included_blocks() {
+                if let Entry::Occupied(mut entry) = self.dependent_messages.entry(block_hash) {
+                    entry.get_mut().remove(&message_id);
+                    if entry.get().is_empty() {
+                        entry.remove_entry();
+                    }
+                }
+            }
+        }
+    }
+
+    fn add_pending_message(&mut self, message: Message, requirements: Vec<B::Hash>) {
+        let message_id = self.next_message_id;
+        self.next_message_id += 1;
+        for block_hash in requirements.iter() {
+            self.dependent_messages
+                .entry(*block_hash)
+                .or_insert_with(HashSet::new)
+                .insert(message_id);
+        }
+        self.message_requirements
+            .insert(message_id, requirements.len());
+        self.pending_messages.insert(message_id, message);
+
+        if message_id >= MESSAGE_ID_BOUNDARY {
+            self.forget_message(message_id - MESSAGE_ID_BOUNDARY)
+        }
+    }
+
+    fn add_message(&mut self, message: Message) {
+        let requirements: Vec<_> = message
+            .included_blocks()
+            .into_iter()
+            .filter(|block_hash| {
+                if self.available_blocks.contains(block_hash) {
+                    return false;
+                }
+                if let Ok(Some(_)) = self.client.header(BlockId::Hash(*block_hash)) {
+                    self.add_block(*block_hash);
+                    return false;
+                }
+                true
+            })
+            .collect();
+
+        if requirements.is_empty() {
+            trace!(target: "afa", "Sending message from DataStore {:?}", message);
+            self.ready_messages_tx
+                .unbounded_send(message)
+                .expect("Member channel should be open");
+        } else {
+            self.add_pending_message(message, requirements);
+        }
+    }
+
+    fn push_messages(&mut self, block_hash: B::Hash) {
+        for message_id in self
+            .dependent_messages
+            .entry(block_hash)
+            .or_insert_with(HashSet::new)
```

Comment:
[Hashmap::remove](https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.remove) returns `Option<V>` so can check that instead of creating unnecessary empty HashSet that you removes in the next block.